resolver/src/p2p/peer/fetcher.rs

```diff
@@ -0,0 +1,222 @@
+use commonware_p2p::{
+    utils::requester::{Requester, ID},
+    Recipients, Sender,
+};
+use commonware_runtime::Clock;
+use commonware_utils::{Array, PrioritySet};
+use governor::clock::Clock as GClock;
+use prost::Message;
+use rand::Rng;
+use std::{
+    marker::PhantomData,
+    time::{Duration, SystemTime},
+};
+use thiserror::Error;
+use tracing::{info, warn};
+
+/// Errors that can occur when sending network messages.
+#[derive(Error, Debug, PartialEq)]
+enum SendError<S: Sender> {
+    #[error("send returned empty")]
+    Empty,
+    #[error("send failed: {0}")]
+    Failed(S::Error),
+}
+
+/// Maintains requests for data from other peers, called fetches.
+///
+/// Requests are called fetches. Fetches may be in one of two states:
+/// - Active: Sent to a peer and is waiting for a response.
+/// - Pending: Not successfully sent to a peer. Waiting to be retried by timeout.
+///
+/// Both types of requests will be retried after a timeout if not resolved (i.e. a response or a
+/// cancelation). Upon retry, requests may either be placed in active or pending state again.
+pub struct Fetcher<
+    E: Clock + GClock + Rng,
+    C: Scheme,
+    Key: Array,
+    NetS: Sender<PublicKey = C::PublicKey>,
+> {
+    context: E,
+
+    /// Helps find peers to fetch from and tracks which peers are assigned to which request ids.
+    requester: Requester<E, C>,
+
+    /// Manages active requests. If a fetch is sent to a peer, it is added to this map.
+    active: BiHashMap<ID, Key>,
+
+    /// Manages pending requests. If fetches fail to make a request to a peer, they are instead
+    /// added to this map and are retried after the deadline.
+    pending: PrioritySet<Key, SystemTime>,
+
+    /// How long fetches remain in the pending queue before being retried
+    retry_timeout: Duration,
+
+    /// Whether requests are sent with priority over other network messages
+    priority_requests: bool,
+
+    /// Phantom data for networking types
+    _s: PhantomData<NetS>,
+}
+
+impl<E: Clock + GClock + Rng, C: Scheme, Key: Array, NetS: Sender<PublicKey = C::PublicKey>>
+    Fetcher<E, C, Key, NetS>
+{
+    /// Creates a new fetcher.
+    pub fn new(
+        context: E,
+        requester: Requester<E, C>,
+        retry_timeout: Duration,
+        priority_requests: bool,
+    ) -> Self {
+        Self {
+            context,
+            requester,
+            active: BiHashMap::new(),
+            pending: PrioritySet::new(),
+            retry_timeout,
+            priority_requests,
+            _s: PhantomData,
+        }
+    }
+
+    /// Makes a fetch request.
+    ///
+    /// If `is_new` is true, the fetch is treated as a new request.
+    /// If false, the fetch is treated as a retry.
+    pub async fn fetch(&mut self, sender: &mut NetS, key: Key, is_new: bool) {
+        // Check if the fetch is already in progress
+        if self.active.contains_right(&key) || self.pending.contains(&key) {
+            info!(?key, "duplicate fetch");
```

Comment:
I think this should be `warn!`. `info` feels like the wrong log level (as it isn't "expected").